### PR TITLE
CMake: Refactor DDR support

### PR DESCRIPTION
**Pull request by dnakamura**
Introduced a number of changes

- Headers can now be directly associated with a ddr set
- A DDR set no longer needs to be specified when enabling DDR for a target.
- Added a new command `ddr_set_add_targets` which adds a target to a DDR
  set. Combined with the above, this allows for a delayed definition of a
  DDR set. In addition targets can be in multiple DDR sets.
- DDR sets can now be composed via `ddr_set_add_targets`. This means we can
  get rid of the awkward `OMR_DDR_SET` mechanics. Instead we just define
  our own DDR set, and consumers simply add our DDR set to theirs

Signed-off-by: Devin Nakamura <devinn@ca.ibm.com>

### CODE REVIEW

**Comment by charliegracie** [comment]
@genie-omr build all

**Comment by charliegracie** [comment]
@keithc-ca thoughts?

File: cmake/modules/OmrDDRSupport.cmake [hunk 1]
```diff
@@ -54,35 +54,68 @@ function(make_ddr_set set_name)
...
-function(target_enable_ddr tgt ddr_set)
+function(target_enable_ddr tgt)
 	if((OMR_HOST_OS STREQUAL "win") OR (NOT OMR_DDR))
 		return()
 	endif()
 
 	set(opt_EARLY_SOURCE_EVAL )
 	set(opt_UNPARSED_ARGUMENTS )
+
+	# Check for people still calling via old style (ie second argumernt is the name of a ddr set)
```
[the remark this answers is not in the document]
Comment (dnakamura) [reply cmake/modules/OmrDDRSupport.cmake]: purely to avoid breaking the openj9 build until i can put changes there

File: cmake/modules/ddr/DDRSetStub.cmake.in [hunk 2]
```diff
@@ -197,16 +216,45 @@ if(DDR_OVERRIDES_FILE)
...
+	${subset_binaries}
+)
+file(WRITE "${DDR_INFO_DIR}/sets/@DDR_TARGET_NAME@.binaries" "${binaries_list}")
+
+
 add_custom_command(
 	OUTPUT ${DDR_BLOB} ${DDR_SUPERSET}
-	DEPENDS macroList
+	DEPENDS ${DDR_MACRO_LIST} ${target_files} ${subset_binaries}
 	COMMAND omr_ddrgen
```
[the remark this answers is not in the document]
Comment (dnakamura) [reply cmake/modules/ddr/DDRSetStub.cmake.in]: Apparently in cmake its called omr_ddrgen, looking into why
https://github.com/eclipse/omr/blob/341d49f5bdd688cc4bbfcfac6cd5170f4d1bad75/ddr/tools/ddrgen/CMakeLists.txt#L22

**Comment by dnakamura** [comment]
@charliegracie  should be good to go

**Comment by charliegracie** [comment]
@genie-omr build all

**Comment by charliegracie** [comment]
@genie-omr build zos

**Comment by charliegracie** [comment]
@keithc-ca were all of your coding concerns addressed?

File: CMakeLists.txt [hunk 3]
```diff
@@ -114,12 +114,12 @@ configure_file(./omrversionstrings.CMakeTemplate.h omrversionstrings.h)
 ###
 include(OmrDDRSupport)
 
-if(NOT DEFINED OMR_DDR_SET)
-	set(OMR_DDR_SET "omr")
-	make_ddr_set(${OMR_DDR_SET})
+make_ddr_set(omrddr)
+# Note  this is a temporary workarround for consumers which still use teh OMR_DDR_SET mechanic
```
Comment (keithc-ca) [inline CMakeLists.txt]: typo: teh -> the; mechanic -> mechanism

File: cmake/modules/OmrDDRSupport.cmake [hunk 4]
```diff
@@ -54,35 +54,69 @@ function(make_ddr_set set_name)
...
+
+
+	cmake_parse_arguments(opt "${options}" "${oneValueArgs}" "${multiValueArgs}"  ${ARGN})
 	omr_assert(FATAL_ERROR TEST NOT opt_UNPARSED_ARGUMENTS MESSAGE "target_enable_ddr: unrecognized options ${opt_UNPARSED_ARGUMENTS}")
 
 
-	set(DDR_SET_TARGET "${ddr_set}_ddr")
-	omr_assert(FATAL_ERROR TEST TARGET ${tgt} MESSAGE "target_enable_ddr called on non-existant target ${tgt}")
-	omr_assert(FATAL_ERROR TEST TARGET "${DDR_SET_TARGET}" MESSAGE "target_enable_ddr called on non-existant ddr_set ${ddr_set}")
+	set(DDR_INFO_DIR "${CMAKE_BINARY_DIR}/ddr_info")
```
Comment (keithc-ca) [inline cmake/modules/OmrDDRSupport.cmake]: I'm not thrilled that `DDR_INFO_DIR` is defined in so many places.

File: cmake/modules/ddr/DDRSetStub.cmake.in [hunk 5]
```diff
@@ -45,6 +54,10 @@ if(NOT DDR_SUPERSET)
 	set(DDR_SUPERSET "${CMAKE_CURRENT_BINARY_DIR}/superset.out")
```
Comment (keithc-ca) [inline cmake/modules/ddr/DDRSetStub.cmake.in]: The default, as defined in `ddrgen.cpp` is for _no_ superset file.

File: cmake/modules/ddr/DDRSetStub.cmake.in [hunk 6]
```diff
@@ -45,6 +54,10 @@ if(NOT DDR_SUPERSET)
...
 endif()
 
+if(NOT DDR_MACRO_LIST)
+	set(DDR_MACRO_LIST "${DDR_INFO_DIR}/sets/@DDR_TARGET_NAME@.macros")
```
Comment (keithc-ca) [inline cmake/modules/ddr/DDRSetStub.cmake.in]: The default, as defined in `ddrgen.cpp` is for no macroList file.

File: cmake/modules/ddr/DDRSetStub.cmake.in [hunk 7]
```diff
@@ -21,22 +21,31 @@
...
+
+project(@DDR_TARGET_NAME@ LANGUAGES NONE)
 
 include("@DDR_TOOLS_EXPORT@")
+include(OmrUtility)
 
-# If not provided, set default values for superset and blob outputs
+# If not provided, set default values for superset, blob, and macrolist outputs
 if(NOT DDR_BLOB)
 	set(DDR_BLOB "${CMAKE_CURRENT_BINARY_DIR}/blob.dat")
```
Comment (keithc-ca) [inline cmake/modules/ddr/DDRSetStub.cmake.in]: The default, as defined in `ddrgen.cpp` is for no binary blob file.

File: cmake/modules/ddr/DDRSetStub.cmake.in [hunk 8]
```diff
@@ -197,16 +215,45 @@ if(DDR_OVERRIDES_FILE)
...
+
 add_custom_command(
 	OUTPUT ${DDR_BLOB} ${DDR_SUPERSET}
-	DEPENDS macroList
-	COMMAND omr_ddrgen
-		${target_files}
+	DEPENDS ${DDR_MACRO_LIST} ${target_files} ${subset_binaries}
+	COMMAND ddrgen
+		${target_files} ${subset_binaries}
 		-e
```
Comment (keithc-ca) [inline cmake/modules/ddr/DDRSetStub.cmake.in]: Can this be changed from `-e` to `--show-empty` for clarity?

File: ddr/lib/ddr-blobgen/CMakeLists.txt [hunk 9]
```diff
@@ -19,21 +19,21 @@
 # SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
 ###############################################################################
 
-add_library(omr_ddr_blobgen
+add_library(ddr_blobgen
```
Comment (keithc-ca) [inline ddr/lib/ddr-blobgen/CMakeLists.txt]: The changes to this file seem unrelated and unnecessary.

File: ddr/lib/ddr-ir/CMakeLists.txt [hunk 10]
```diff
@@ -19,7 +19,7 @@
 # SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
 ###############################################################################
 
-add_library(omr_ddr_ir
+add_library(ddr_ir
```
Comment (keithc-ca) [inline ddr/lib/ddr-ir/CMakeLists.txt]: The changes to this file seem unrelated and unnecessary.

File: ddr/lib/ddr-macros/CMakeLists.txt [hunk 11]
```diff
@@ -19,14 +19,14 @@
 # SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
 ###############################################################################
 
-add_library(omr_ddr_macros
+add_library(ddr_macros
```
Comment (keithc-ca) [inline ddr/lib/ddr-macros/CMakeLists.txt]: The changes to this file seem unrelated and unnecessary.

File: ddr/lib/ddr-scanner/CMakeLists.txt [hunk 12]
```diff
@@ -19,70 +19,70 @@
 # SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
 ###############################################################################
 
-add_library(omr_ddr_scanner
+add_library(ddr_scanner
```
Comment (keithc-ca) [inline ddr/lib/ddr-scanner/CMakeLists.txt]: The changes to this file seem unrelated and unnecessary.

File: ddr/tools/CMakeLists.txt [hunk 13]
```diff
@@ -22,4 +22,4 @@
 add_subdirectory(blob_reader)
 add_subdirectory(ddrgen)
 
-export(TARGETS omr_blob_reader omr_ddrgen FILE "DDRTools.cmake")
+export(TARGETS blob_reader ddrgen FILE "DDRTools.cmake")
```
Comment (keithc-ca) [inline ddr/tools/CMakeLists.txt]: The changes to this file seem unrelated and unnecessary.

File: ddr/tools/blob_reader/CMakeLists.txt [hunk 14]
```diff
@@ -19,17 +19,17 @@
 # SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
 ###############################################################################
 
-add_executable(omr_blob_reader
+add_executable(blob_reader
```
Comment (keithc-ca) [inline ddr/tools/blob_reader/CMakeLists.txt]: The changes to this file seem unrelated and unnecessary.

File: ddr/tools/ddrgen/CMakeLists.txt [hunk 15]
```diff
@@ -19,25 +19,25 @@
 # SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
 ###############################################################################
 
-add_executable(omr_ddrgen
+add_executable(ddrgen
```
Comment (keithc-ca) [inline ddr/tools/ddrgen/CMakeLists.txt]: The changes to this file seem unrelated and unnecessary.

File: ddr/lib/ddr-macros/CMakeLists.txt (hunk 11 above)
Comment (dnakamura) [reply ddr/lib/ddr-macros/CMakeLists.txt]: these changes are to bring things inline with the makefile target names (which do not have the `omr_` prefix

Comment (keithc-ca) [reply ddr/lib/ddr-macros/CMakeLists.txt]: I think these changes would have been better isolated in a separate pull request.

**Comment by dnakamura** [comment]
Should all be resolved

File: cmake/modules/ddr/DDRSetStub.cmake.in [hunk 16]
```diff
@@ -197,16 +211,31 @@ if(DDR_OVERRIDES_FILE)
 	list(APPEND EXTRA_DDRGEN_OPTIONS "--overrides" "${DDR_OVERRIDES_FILE}" WORKING_DIRECTORY "${override_dir}")
```
Comment (keithc-ca) [inline cmake/modules/ddr/DDRSetStub.cmake.in]: I think `WORKING_DIRECTORY` should be specified independent of whether `DDR_OVERRIDES_FILE` is set, for example, the blacklist may be or other file names may be relative.

File: cmake/modules/ddr/DDRSetStub.cmake.in (hunk 8 above)
Comment (keithc-ca) [reply cmake/modules/ddr/DDRSetStub.cmake.in]: @dnakamura ?

File: cmake/modules/ddr/DDRSetStub.cmake.in (hunk 16 above)
Comment (dnakamura) [reply cmake/modules/ddr/DDRSetStub.cmake.in]: THe issue I have with that is that it potentially changes behavior, when you specify the overrides file (the overrides / blacklist etc may reside in different directories). Instead, i have changed it so that relative file paths are all interred as relative to the directory where the ddr_set is declared

**Comment by dnakamura** [comment]
@keithc-ca  all good?

File: cmake/modules/ddr/DDRSetStub.cmake.in (hunk 16 above)
Comment (keithc-ca) [reply cmake/modules/ddr/DDRSetStub.cmake.in]: I think it is reasonable that relative paths be interpreted based on the ddr_set directory. That suggest the name `override_dir` could be improved.

File: cmake/modules/ddr/DDRSetStub.cmake.in (hunk 8 above)
Comment (keithc-ca) [reply cmake/modules/ddr/DDRSetStub.cmake.in]: I'd like to see the longer option be used here.

**Comment by dnakamura** [comment]
should all be fixed

**Comment by charliegracie** [comment]
@genie-omr build all

File: cmake/modules/OmrDDRSupport.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (keithc-ca) [reply cmake/modules/OmrDDRSupport.cmake]: > @keithc-ca all good?

No, many of my comments seem to have been ignored.

**Comment by keithc-ca** [comment]
> @keithc-ca all good?

No, many of my comments seem to have been ignored.

**Comment by dnakamura** [comment]
@keithc-ca  can you point to specific comments / mark them unresolved? Looking through the comments I cant see anything that has not been addressed

File: CMakeLists.txt (hunk 3 above)
Comment (keithc-ca) [reply CMakeLists.txt]: These typos are not fixed; also workarround -> workaround.

Comment (dnakamura) [reply CMakeLists.txt]: This is an outdated diff, it is not in the final PR

File: CMakeLists.txt [hunk 17]
```diff
@@ -114,12 +114,12 @@ configure_file(./omrversionstrings.CMakeTemplate.h omrversionstrings.h)
 ###
 include(OmrDDRSupport)
 
-if(NOT DEFINED OMR_DDR_SET)
-	set(OMR_DDR_SET "omr")
-	make_ddr_set(${OMR_DDR_SET})
+make_ddr_set(omrddr)
+# Note  this is a temporary workarround for consumers which still use teh OMR_DDR_SET mechanism
```
Comment (keithc-ca) [inline CMakeLists.txt]: My comment relates to this (not outdated) commit.

Comment (dnakamura) [reply CMakeLists.txt]: removed, since the only one this would help is openj9, which does not have fully functional ddr support from cmake anyway

**Comment by youngar** [comment]
@genie-omr build all

**Comment by dnakamura** [comment]
@keithc-ca ?

**Comment by keithc-ca** [comment]
@dnakamura I believe you have addressed all my comments.
